src/datahike/db.cljc

```diff
@@ -1058,24 +600,38 @@
              (is-attr? db (.-a datom) :db/unique))
     (when-let [found (not-empty (-datoms db :avet [(.-a datom) (.-v datom)]))]
       (raise "Cannot add " datom " because of unique constraint: " found
-             {:error :transact/unique
+             {:error     :transact/unique
               :attribute (.-a datom)
-              :datom datom}))))
+              :datom     datom}))))
 
 (defn- validate-eid [eid at]
   (when-not (number? eid)
     (raise "Bad entity id " eid " at " at ", expected number"
            {:error :transact/syntax, :entity-id eid, :context at})))
 
-(defn- validate-attr [attr at]
-  (when-not (or (keyword? attr) (string? attr))
-    (raise "Bad entity attribute " attr " at " at ", expected keyword or string"
-           {:error :transact/syntax, :attribute attr, :context at})))
-
-(defn- validate-val [v at]
+(defn- validate-attr [attr at db]
+  (if (get-in db [:config :schema-on-read])
+    (when-not (or (keyword? attr) (string? attr))
+      (raise "Bad entity attribute " attr " at " at ", expected keyword or string"
+             {:error :transact/syntax, :attribute attr, :context at}))
+    (when-not (ds/schema-attr? attr)
+      (if-let [db-idents (-> db :rschema :db/ident)]
+        (when-not (db-idents attr)
+          (raise "Bad entity attribute " attr " at " at ", not defined in current schema"
+                 {:error :transact/schema :attribute attr :context at}))
+        (raise "No schema found in db."
+               {:error :transact/schema :attribute attr :context at})))))
+
+(defn- validate-val [v [_ _ a _ _ :as at] db]
   (when (nil? v)
     (raise "Cannot store nil as a value at " at
-           {:error :transact/syntax, :value v, :context at})))
+           {:error :transact/syntax, :value v, :context at}))
+  (when-not (get-in db [:config :schema-on-read])
+    (let [schema (:schema db)
+          schema-spec (if (ds/schema-attr? a) ds/implicit-schema-spec schema)]
+      (when-not (ds/value-valid? at schema)
+        (raise "Bad entity value " v " at " at ", value does not match schema definition. Must be conform to: " (ds/describe-type (get-in schema-spec [a :db/valueType]))
```

Comment:
There is duplication between the error message creation here and in the test. May be the string creation should be moved into a function?